### ISSUE TITLE
How to Use Both HTTP and HTTPS with Custom Ports and HTML Content Protected by Let's Encrypt SSL Certificates?

### ISSUE THREAD

**Issue by own3mall** [fiorix/freegeoip#228]
According to the documentation from the docker package, shouldn't I be able to:

Listen to HTTP requests on all IPs on port 13333
Listen to HTTPS requests on all IPs on port 13334
Serve static HTML content from the "/var/www/geoservices" directory if no API format is specified (basically the "/" page)?  (does it try to load index.html by default?)
Use SSL certificates provided by Let's Encrypt for the hosts of "www.domainhere.com,domain.com"

Using this command:

```
docker run --restart=always -p 13333:13334 -d fiorix/freegeoip -http 0.0.0.0:13333 -https 0.0.0.0:13334 -public /var/www/geoservices -letsencrypt -letsencrypt-email avalidemail@avalidemail.com -letsencrypt-hosts www.domainhere.com,domain.com
```

What am I doing wrong?  The documentation doesn't seem to go into this much detail.  Using the above command results in neither HTTP or HTTPS working.

Any help is appreciated.

**Comment by fiorix** [comment]
The freegeoip -http and -https parameters are correct, but your docker -p parameter is not:

https://docs.docker.com/engine/reference/run/#expose-incoming-ports

You probably want docker -p 13333:13333 -p 13334:13334 to have both ports 13333 and 13334 exposed on the host machine running docker, mapping to their counterparts inside the container.

**Comment by own3mall** [comment]
Thanks, but I still can't get it work with that change:

Chrome:

> This site can’t provide a secure connection
> 
> {domain} sent an invalid response.
> Try running Windows Network Diagnostics.
> ERR_SSL_PROTOCOL_ERROR

Firefox:

> Secure Connection Failed
> 
> An error occurred during a connection to {domain}. Peer reports it experienced an internal error. Error code: SSL_ERROR_INTERNAL_ERROR_ALERT
> 
> The page you are trying to view cannot be shown because the authenticity of the received data could not be verified.

Do the strings have to quoted for Let's Encrypt to work in the command in my first post?

Also, is there any documentation on what and how the `-public` flag works?

**Comment by fiorix** [comment]
I don't know what could be wrong with the network part, but if you look for logs from the docker console there should be something.

The `-public` flag is pretty straightforward. If set, it exposed the directory point to it as the `/` endpoint in the server.

**Comment by own3mall** [comment]
Yeah, I'm not sure what's going on.  Thanks a lot for your help so far though!

I am now running this command, but I cannot get Let's Encrypt to work:

```
docker run --restart=always -p 9999:9999 -p 10000:10000 -d fiorix/freegeoip -http 0.0.0.0:9999 -https 0.0.0.0:10000 -public /var/www/geoservices -letsencrypt -letsencrypt-email MYEMAIL -letsencrypt-hosts www.MYDOMAIN.com,MYDOMAIN.com
```

What pre-requisites are required for Let's Encrypt integration?  I have certbot installed manually for other software I use (won't conflict with freegeoip) and it runs correctly, meaning certbot resolved its dependencies.  

Also, the `-public` flag does not appear to work.   In the example above, `/var/www/geoservices/index.html` does not load when using the `-public /var/www/geoservices` parameter.  I'm probably using it incorrectly though.

This is what I have in the docker logs.

```
2017/07/14 03:05:42 freegeoip https server starting on 0.0.0.0:10000
2017/07/14 03:05:42 database info: starting update
2017/07/14 03:05:42 freegeoip http server starting on 0.0.0.0:9999
2017/07/14 03:05:42 database info: finished update
2017/07/14 03:05:44 database loaded: /tmp/freegeoip/db.gz
[access] 24.8.108.66 - - [14/Jul/2017:03:05:47 +0000] "GET / HTTP/1.1" 404 19 ""                                                    "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:45.9) Gecko/20100101 Goanna/3.2 Fi                                                   refox/45.9 PaleMoon/27.4.0"
[access] 24.8.108.66 - - [14/Jul/2017:03:05:47 +0000] "GET /favicon.ico HTTP/1.1                                                   " 404 19 "" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:45.9) Gecko/20100101 Go                                                   anna/3.2 Firefox/45.9 PaleMoon/27.4.0"
[access] 24.8.108.66 - - [14/Jul/2017:03:05:47 +0000] "GET /favicon.ico HTTP/1.1                                                   " 404 19 "" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:45.9) Gecko/20100101 Go                                                   anna/3.2 Firefox/45.9 PaleMoon/27.4.0"
[access] 24.8.108.66 - - [14/Jul/2017:03:05:47 +0000] "GET /favicon.ico HTTP/1.1                                                   " 404 19 "" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:45.9) Gecko/20100101 Go                                                   anna/3.2 Firefox/45.9 PaleMoon/27.4.0"
[access] 24.8.108.66 - - [14/Jul/2017:03:05:47 +0000] "GET /favicon.ico HTTP/1.1                                                   " 404 19 "" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:45.9) Gecko/20100101 Go                                                   anna/3.2 Firefox/45.9 PaleMoon/27.4.0"
[error] 2017/07/14 03:05:57 http: TLS handshake error from 24.8.108.66:2730: acm                                                   e: identifier authorization failed
[error] 2017/07/14 03:05:57 http: TLS handshake error from 24.8.108.66:2731: acm                                                   e/autocert: missing certificate
[access] 24.8.108.66 - - [14/Jul/2017:03:05:57 +0000] "GET /favicon.ico HTTP/1.1                                                   " 404 19 "" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:45.9) Gecko/20100101 Go                                                   anna/3.2 Firefox/45.9 PaleMoon/27.4.0"
[error] 2017/07/14 03:06:12 http: TLS handshake error from 24.8.108.66:2742: acm                                                   e/autocert: missing certificate
[error] 2017/07/14 03:06:26 http: TLS handshake error from 24.8.108.66:2783: acm                                                   e/autocert: missing certificate
[error] 2017/07/14 03:06:26 http: TLS handshake error from 24.8.108.66:2784: acm                                                   e/autocert: missing certificate
[error] 2017/07/14 03:06:30 http: TLS handshake error from 24.8.108.66:2786: acm                                                   e/autocert: missing certificate
[access] 24.8.108.66 - - [14/Jul/2017:03:09:51 +0000] "GET / HTTP/1.1" 404 19 ""                                                    "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:54.0) Gecko/20100101 Firefox/54.0"
[access] 24.8.108.66 - - [14/Jul/2017:03:09:52 +0000] "GET /favicon.ico HTTP/1.1                                                   " 404 19 "" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:54.0) Gecko/20100101 Firefox                                                   /54.0"
[access] 24.8.108.66 - - [14/Jul/2017:03:09:52 +0000] "GET /favicon.ico HTTP/1.1                                                   " 404 19 "" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:54.0) Gecko/20100101 Firefox                                                   /54.0"
[access] 24.8.108.66 - - [14/Jul/2017:03:09:57 +0000] "GET /json/ HTTP/1.1" 200                                                    247 "" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:54.0) Gecko/20100101 Firefox/54.0                                                   "
2017/07/15 03:05:42 database info: starting update
2017/07/15 03:05:43 database info: finished update
2017/07/16 03:05:43 database info: starting update
2017/07/16 03:05:43 database info: finished update
[access] 95.215.61.97 - - [16/Jul/2017:04:06:47 +0000] "GET / HTTP/1.1" 404 19 "                                                   " "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:52.0) Gecko/20100101 Firefox/52.0"
[access] 66.249.69.70 - - [16/Jul/2017:17:38:52 +0000] "GET /robots.txt HTTP/1.1                                                   " 404 19 "" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.                                                   html)"
[access] 66.249.69.66 - - [16/Jul/2017:17:38:53 +0000] "GET / HTTP/1.1" 404 19 "                                                   " "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
2017/07/17 03:05:43 database info: starting update
2017/07/17 03:05:44 database info: finished update
[error] 2017/07/17 09:43:32 http: TLS handshake error from 5.188.11.103:57526: t                                                   ls: first record does not look like a TLS handshake
2017/07/18 03:05:44 database info: starting update
2017/07/18 03:05:44 database info: finished update
2017/07/19 03:05:44 database info: starting update
2017/07/19 03:05:45 database info: finished update
[access] 50.112.184.15 - - [19/Jul/2017:04:27:09 +0000] "GET / HTTP/1.1" 404 19                                                    "" "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:31.0) Gecko/20100101 Firefox/31.0"
[access] 66.249.64.69 - - [19/Jul/2017:17:38:54 +0000] "GET /robots.txt HTTP/1.1                                                   " 404 19 "" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.                                                   html)"
[access] 66.249.64.66 - - [19/Jul/2017:17:38:55 +0000] "GET / HTTP/1.1" 404 19 "                                                   " "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
[error] 2017/07/19 23:05:04 http: TLS handshake error from 164.52.0.132:36445: a                                                   cme/autocert: missing server name
[error] 2017/07/19 23:05:07 http: TLS handshake error from 164.52.0.132:40106: r                                                   ead tcp 172.17.0.2:10000->164.52.0.132:40106: read: connection reset by peer
[error] 2017/07/19 23:05:07 http: TLS handshake error from 164.52.0.132:54864: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:08 http: TLS handshake error from 164.52.0.132:55733: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:08 http: TLS handshake error from 164.52.0.132:48900: t                                                   ls: oversized record received with length 55670
[error] 2017/07/19 23:05:09 http: TLS handshake error from 164.52.0.132:33664: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:09 http: TLS handshake error from 164.52.0.132:36191: t                                                   ls: oversized record received with length 51456
[error] 2017/07/19 23:05:09 http: TLS handshake error from 164.52.0.132:38463: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:10 http: TLS handshake error from 164.52.0.132:42875: t                                                   ls: oversized record received with length 18767
[error] 2017/07/19 23:05:10 http: TLS handshake error from 164.52.0.132:55397: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:11 http: TLS handshake error from 164.52.0.132:50147: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:11 http: TLS handshake error from 164.52.0.132:60900: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:12 http: TLS handshake error from 164.52.0.132:56770: t                                                   ls: oversized record received with length 29811
[error] 2017/07/19 23:05:12 http: TLS handshake error from 164.52.0.132:36712: t                                                   ls: oversized record received with length 28532
[error] 2017/07/19 23:05:12 http: TLS handshake error from 164.52.0.132:60467: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:13 http: TLS handshake error from 164.52.0.132:45798: t                                                   ls: oversized record received with length 23296
[error] 2017/07/19 23:05:13 http: TLS handshake error from 164.52.0.132:50622: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:14 http: TLS handshake error from 164.52.0.132:35329: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:14 http: TLS handshake error from 164.52.0.132:60737: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:15 http: TLS handshake error from 164.52.0.132:59352: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:15 http: TLS handshake error from 164.52.0.132:56970: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:15 http: TLS handshake error from 164.52.0.132:60970: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:16 http: TLS handshake error from 164.52.0.132:59215: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:16 http: TLS handshake error from 164.52.0.132:47860: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:20 http: TLS handshake error from 164.52.0.132:45245: t                                                   ls: oversized record received with length 21024
[error] 2017/07/19 23:05:20 http: TLS handshake error from 164.52.0.132:54105: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:20 http: TLS handshake error from 164.52.0.132:39478: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:21 http: TLS handshake error from 164.52.0.132:47392: t                                                   ls: oversized record received with length 22862
[error] 2017/07/19 23:05:21 http: TLS handshake error from 164.52.0.132:58505: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:25 http: TLS handshake error from 164.52.0.132:47824: r                                                   ead tcp 172.17.0.2:10000->164.52.0.132:47824: read: connection reset by peer
[error] 2017/07/19 23:05:25 http: TLS handshake error from 164.52.0.132:54633: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:26 http: TLS handshake error from 164.52.0.132:48625: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:26 http: TLS handshake error from 164.52.0.132:32992: t                                                   ls: oversized record received with length 20256
[error] 2017/07/19 23:05:30 http: TLS handshake error from 164.52.0.132:34686: t                                                   ls: oversized record received with length 18688
[error] 2017/07/19 23:05:30 http: TLS handshake error from 164.52.0.132:44225: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:30 http: TLS handshake error from 164.52.0.132:52703: t                                                   ls: oversized record received with length 21504
[error] 2017/07/19 23:05:31 http: TLS handshake error from 164.52.0.132:58384: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:31 http: TLS handshake error from 164.52.0.132:44358: t                                                   ls: oversized record received with length 21248
[error] 2017/07/19 23:05:32 http: TLS handshake error from 164.52.0.132:47256: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:32 http: TLS handshake error from 164.52.0.132:42649: t                                                   ls: first record does not look like a TLS handshake
[error] 2017/07/19 23:05:32 http: TLS handshake error from 164.52.0.132:43805: t                                                   ls: oversized record received with length 28012
[error] 2017/07/19 23:05:33 http: TLS handshake error from 164.52.0.132:41344: t                                                   ls: oversized record received with length 26990
2017/07/20 03:05:45 database info: starting update
2017/07/20 03:05:46 database info: finished update
[error] 2017/07/20 05:12:07 http: TLS handshake error from 80.82.77.33:53072: ac                                                   me/autocert: host not configured
[error] 2017/07/20 05:12:08 http: TLS handshake error from 80.82.77.33:53715: tl                                                   s: first record does not look like a TLS handshake
```

Not very helpful unfortunately.  I don't see anything regarding Let's Encrypt actually running.  😞

**Comment by fiorix** [comment]
Hey, seems like you're having different problems, so let's try one by one.

1. Port mapping in docker sets the local port first, then the port inside the container
2. The directory you're trying to set to -public does not exist in the container, you have to map it
3. Let's encrypt is failing due to authorization problems, you have it in your logs

When I run, for example, this:

```
ubuntu@ubuntu-xenial:~$ mkdir my-www
ubuntu@ubuntu-xenial:~$ echo hello world > my-www/index.html
ubuntu@ubuntu-xenial:~$ docker run -p 8080:80 -p 8443:443 -v `pwd`/my-www:/tmp/www -i -t fiorix/freegeoip -http :80 -https :443 -public /tmp/www -letsencrypt -letsencrypt-hosts localhost
2017/07/22 18:15:36 freegeoip https server starting on :443
2017/07/22 18:15:36 database info: starting update
2017/07/22 18:15:36 freegeoip http server starting on :80
2017/07/22 18:15:42 database info: finished update
2017/07/22 18:15:43 database loaded: /tmp/freegeoip/db.gz
```

I can go on another terminal and hit the public dir:

```
ubuntu@ubuntu-xenial:~$ curl localhost:8080
hello world
```

But I cannot use letsencrypt with localhost:

```
ubuntu@ubuntu-xenial:~$ curl https://localhost:8443
curl: (35) gnutls_handshake() failed: Internal error
```

Server dumps this:

```
[error] 2017/07/22 18:16:45 http: TLS handshake error from 172.17.0.1:59646: acme/autocert: server name component count invalid
```

So basically if you have your port mappings and directory mappings correct, and your lets encrypt settings correct, it should work as expected.

**Comment by own3mall** [comment]
Thanks, that is getting me closer and helping quite a bit.  Is there a way to serve a single static HTML page as oppose to showing a listing of files at the root index of the directory?

**Comment by fiorix** [comment]
By default it'll look for an index.html. You can create an empty file to avoid directory listing.

**Comment by own3mall** [comment]
Thanks, but no matter what I try, I cannot get HTTPS to work.  Let's Encrypt parameters don't seem to do anything.

```
docker run --restart=always -p 10001:10001 -p 10002:10002 -v /var/www/dir:/tmp/www -i -t fiorix/freegeoip -http :10001 -https :10002 -public /tmp/www -letsencrypt -letsencrypt-email myemail@gmail.com -letsencrypt-hosts www.host.com,host.com
```

There doesn't appear to be any logging in terms of Let's Encrypt, so does this functionality even work?

```
[error] 2017/08/18 02:00:15 http: TLS handshake error from 67.165.238.231:2870: acme: identifier authorization failed
[error] 2017/08/18 02:00:15 http: TLS handshake error from 67.165.238.231:2872: acme/autocert: missing certificate
[access] 67.165.238.231 - - [18/Aug/2017:02:00:28 +0000] "GET /json/ HTTP/1.1" 200 243 "" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:45.9) Gecko/20100101 Goanna/3.2 Firefox/45.9 PaleMoon/27.4.1"
```

**Comment by MACscr** [comment]
Did you have any luck getting this to work?

**Comment by fiorix** [comment]
It does work, I just tested on my personal server. It's possible that letsencrypt wants your server to be running on port 443, not 10002.

**Comment by own3mall** [comment]
I'll try it on port 443 and see if that makes a difference.

Thanks.

**Comment by own3mall** [comment]
Wondering if you can use a custom port with Let's Encrypt...

-p 8443:443 work since the container thinks it's on 443?

**Comment by MACscr** [comment]
@fiorix. Seems using Letsencrypt with such a setup is pretty common and something maybe you have already done. Could you by chance provide us with a repeatable process that we can use ourselves to make this happen? The readme for using SSL with your docker image seems a bit vague. With that said, thank you such for your time on this project!

**Comment by fiorix** [comment]
Hey I'm happy to update the documentation with whatever suggestions you have. Even better if you can send a PR with the updates for me to review.

To clarify all this, I just tested again on a brand new droplet @ digital ocean, configure DNS pointing to it, and started the server as follows:

```
root@ubuntu-512mb-lon1-01:~# host freegeoip.fiorix.net
freegeoip.fiorix.net has address 138.68.158.184
root@ubuntu-512mb-lon1-01:~# docker run --net=host -i -t fiorix/freegeoip -https=:443 -letsencrypt -letsencrypt-hosts=freegeoip.fiorix.net
Unable to find image 'fiorix/freegeoip:latest' locally
latest: Pulling from fiorix/freegeoip
9f0706ba7422: Pull complete
d3942a742d22: Pull complete
62b1123c88f6: Pull complete
3306e13140ef: Pull complete
f0a19cb07fbd: Pull complete
8dfc5cc1a7eb: Pull complete
9e13dd569e2c: Pull complete
30a138bed8a3: Pull complete
c2ba2577ca54: Pull complete
179fec31ec22: Pull complete
Digest: sha256:223cd7eaeed4e3b8865844d6a71e4d1b048aefed6b326a1c65787a2f59c37f0a
Status: Downloaded newer image for fiorix/freegeoip:latest
2017/09/01 10:13:57 freegeoip https server starting on :443
2017/09/01 10:13:57 database info: starting update
2017/09/01 10:13:57 freegeoip http server starting on :8080
2017/09/01 10:13:57 database info: finished update
2017/09/01 10:13:59 database loaded: /tmp/freegeoip/db.gz
```

Then from another terminal:

```
~ $ curl https://freegeoip.fiorix.net/json/github.com
{"ip":"192.30.253.112","country_code":"US","country_name":"United States","region_code":"CA","region_name":"California","city":"San Francisco","zip_code":"94107","time_zone":"America/Los_Angeles","latitude":37.7697,"longitude":-122.3933,"metro_code":807}
```

Obviously you wouldn't want to use `--net=host` in prod with docker, but it serves as a confirmation that letsencrypt works as expected even from docker. To your other question, I don't think letsencrypt supports any other port than 443.

**Comment by fiorix** [comment]
Destroying that droplet now.

**Comment by own3mall** [comment]
-p 8443:443 but using 443 for everything as you said does work.

Thanks.  I wish Let's Encrypt hadn't put this requirement on us...

**Comment by danielpost** [comment]
@own3mall Can you give an example of your command? I'm using `docker run --restart=always -p 80:80 -p 443:443 -d fiorix/freegeoip -http :80 -https :443 -letsencrypt -letsencrypt-hosts=subdomain.example.com` and it works over HTTP but it doesn't seem to do anything related to SSL.

**Comment by own3mall** [comment]
@danielpost , sorry for the late response, but I use a variant of this for the command:

```
docker run --restart=always -p 4555:4555 -p 443:443 -v /var/www/geoservices:/tmp/www -d fiorix/freegeoip -http :4555 -https :443 -public /tmp/www -letsencrypt -letsencrypt-email myemailaddress@gmail.com -letsencrypt-hosts www.mysite.com,mysite.com
```
